### PR TITLE
*: rewrite update-ssh-keys in rust

### PR DESCRIPTION
**Pull request by sdemos**
This replaces both the authorized_keys.d library and the command line tool.

### CODE REVIEW

**Comment by crawford** [comment]
Travis config?

**Comment by lucab** [comment]
It looks like ignition has a reference to this package in its glide.yaml. Should we maybe first move the existing go package over there?

**Comment by sdemos** [comment]
I was thinking that we were just going to continue to refer to the old commit in the tree, but talking to derek, he pointed out that then we wouldn't be able to patch it if we find a bug. Instead, since ignition is the only other consumer of the library portion of this project, I opened a pr to move it under there (coreos/ignition#467)

File: .gitignore [hunk 1]
```diff
@@ -1,3 +1,3 @@
-bin/
-gopath/
-*.swp
+/target/
+**/*.rs.bk
+Cargo.lock
```
Comment (euank) [inline .gitignore]: Why aren't we checking in `Cargo.lock`?

Comment (sdemos) [reply .gitignore]: we can. iirc this is just how cargo set it up when I started the project. maybe that's one of the `cargo new --bin` vs `cargo new --lib` differences.

Comment (euank) [reply .gitignore]: It is; `bin` leaves it unignored.

Isn't this a `bin` not a `lib` really though?

**Comment by bgilbert** [comment]
Have you checked that coreos-cloudinit is happy with this implementation?

**Comment by sdemos** [comment]
I haven't tested it with this implementation, but cloudinit calls the binary [here](https://github.com/coreos/coreos-cloudinit/blob/3c68e2e67d2ea077e2b79383c0ad313770c62ca2/system/ssh_key.go#L36) using a configuration of flags that I've tested and it works fine. it just runs in userspace like a normal program right?

**Comment by bgilbert** [comment]
@sdemos It does.  We should still specifically test running it from coreos-cloudinit at boot, though, in case there's breakage caused by a difference in environment variables, etc.

File: .gitignore (hunk 1 above)
Comment (sdemos) [reply .gitignore]: it's a little of both. it's the command line utility but also a library consumed by `coreos-metadata`. I can include it.

File: src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,400 @@
...
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! update-ssh-keys library
+//!
+//! this library is meant to replace github.com/coreos/update-ssh-keys
```
Comment (dm0-) [inline src/lib.rs]: This line will be self-referential when merged, which will look odd.

**Comment by sdemos** [comment]
I tested provisioning a machine with coreos-cloudinit with this binary in the image instead of the shell script and it worked as expected.

**Review by crawford — APPROVED** [review]

**Pull request merged**